core/common/src/main/java/alluxio/util/logging/SamplingLogger.java

```diff
@@ -0,0 +1,453 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.util.logging;
+
+import org.slf4j.Logger;
+import org.slf4j.Marker;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A logger which limits log message frequency to prevent spamming the logs.
+ *
+ * After the logger logs a message, it enters a configurable cooldown period where it will ignore
+ * repeated logging requests with the same message. Messages are compared by message template, not
+ * message arguments.
+ */
+public class SamplingLogger implements Logger {
+  private final Logger mDelegate;
+  private final long mCooldownMs;
+  private final Map<String, AtomicLong> mCooldowns = new ConcurrentHashMap<>();
```

Comment:
this could grow forever, right? Should old entries get aged out?

Reply:
This matches by format string, so the size of this map shouldn't exceed the number of lines of code where we log, assuming we are writing our log messages properly (`LOG.info("{} did {}", arg1, arg2")` instead of `LOG.info(arg1 + " did " + arg2)`)